sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanIterators.scala

```diff
@@ -0,0 +1,317 @@
+
+  override def hasNext: Boolean = {
+    // Kill the task in case it has been marked as killed. This logic is from
+    // InterruptibleIterator, but we inline it here instead of wrapping the iterator in order
+    // to avoid performance overhead.
+    context.killTaskIfInterrupted()
+    (currentIterator != null && currentIterator.hasNext) || nextIterator()
+  }
+
+  override def next(): Object
+
+  private def readFile(file: PartitionedFile): Iterator[InternalRow] = {
+    try {
+      readFunction(file)
+    } catch {
+      case e: FileNotFoundException =>
+        throw new FileNotFoundException(
+          e.getMessage + "\n" +
+            "It is possible the underlying files have been updated. " +
+            "You can explicitly invalidate the cache in Spark by " +
+            "running 'REFRESH TABLE tableName' command in SQL or " +
+            "by recreating the Dataset/DataFrame involved.")
+    }
+  }
+
+  /** Advances to the next file. Returns true if a new non-empty iterator is available. */
+  protected def nextIterator(): Boolean = {
+    if (files.hasNext) {
+      currentFile = files.next()
+      logInfo(s"Reading File $currentFile")
+      // Sets InputFileBlockHolder for the file block's information
+      InputFileBlockHolder.set(currentFile.filePath, currentFile.start, currentFile.length)
+
+      if (ignoreMissingFiles || ignoreCorruptFiles) {
+        currentIterator = new NextIterator[Object] {
+          private val file = currentFile
+          // The readFunction may read some bytes before consuming the iterator, e.g.,
+          // vectorized Parquet reader. Here we use lazy val to delay the creation of
+          // iterator so that we will throw exception in `getNext`.
+          private lazy val internalIter = readFile(file)
+
+          override def getNext(): AnyRef = {
+            try {
+              if (internalIter.hasNext) {
+                internalIter.next()
+              } else {
+                finished = true
+                null
+              }
+            } catch {
+              case e: FileNotFoundException if ignoreMissingFiles =>
+                logWarning(s"Skipped missing file: $currentFile", e)
+                finished = true
+                null
+              // Throw FileNotFoundException even if `ignoreCorruptFiles` is true
+              case e: FileNotFoundException if !ignoreMissingFiles => throw e
+              case e @ (_: RuntimeException | _: IOException) if ignoreCorruptFiles =>
+                logWarning(
+                  s"Skipped the rest of the content in the corrupted file: $currentFile", e)
+                finished = true
+                null
+            }
+          }
+
+          override def close(): Unit = {}
+        }
+      } else {
+        currentIterator = readFile(currentFile)
+      }
+
+      try {
+        context.killTaskIfInterrupted()
+        (currentIterator != null && currentIterator.hasNext) || nextIterator()
```

Comment:
This is changed compared to `FileScanRDD.nextIterator`, not call `hasNext()` because `FileSortedBucketScanIterator` needs to override `hasNext()`.